File: 6/receiver_imap/main.go
package main

import (
	"encoding/json"
	"fmt"
	"io"
	"io/ioutil"
	"log"
	"os"

	"github.com/emersion/go-imap"
	"github.com/emersion/go-imap/client"
)

type Config struct {
	SMTPHost      string `json:"smtpHost"`
	SMTPPort      string `json:"smtpPort"`
	IMAPServer    string `json:"imapServer"`
	Username      string `json:"username"`
	Password      string `json:"password"`
	From          string `json:"from"`
	To            string `json:"to"`
	LabName       string `json:"labName"`
	SenderName    string `json:"senderName"`
	SenderGroup   string `json:"senderGroup"`
	ReceiverName  string `json:"receiverName"`
	ReceiverGroup string `json:"receiverGroup"`
}

func loadConfig(filename string) (*Config, error) {
	data, err := ioutil.ReadFile(filename)
	if err != nil {
		return nil, err
	}
	var config Config
	if err := json.Unmarshal(data, &config); err != nil {
		return nil, err
	}
	return &config, nil
}

func fetchEmails(config *Config) error {
	c, err := client.DialTLS(config.IMAPServer, nil)
	if err != nil {
		return err
	}
	defer c.Logout()

	if err := c.Login(config.Username, config.Password); err != nil {
		return err
	}

	mbox, err := c.Select("INBOX", false)
	if err != nil {
		return err
	}

	fmt.Printf("Количество писем в ящике: %d\n", mbox.Messages)

	seqset := new(imap.SeqSet)
	if mbox.Messages > 0 {
		var fromSeq uint32 = 1
		if mbox.Messages > 5 {
			fromSeq = mbox.Messages - 5 + 1
		}
		seqset.AddRange(fromSeq, mbox.Messages)
	}

	messages := make(chan *imap.Message, 5)
	go func() {
		if err := c.Fetch(seqset, []imap.FetchItem{imap.FetchEnvelope}, messages); err != nil {
			log.Fatal(err)
		}
	}()

	fmt.Println("\nЗаголовки последних писем:")
	count := 0
	for msg := range messages {
		if msg.Envelope != nil {
			fmt.Printf("От: %v, Тема: %v, Дата: %v\n", msg.Envelope.From, msg.Envelope.Subject, msg.Envelope.Date)
		}
		count++
		if count >= 5 {
			break
		}
	}

	criteria := imap.NewSearchCriteria()
	criteria.Header.Add("From", config.To)
	uids, err := c.Search(criteria)
	if err != nil {
		return err
	}

	if len(uids) == 0 {
		fmt.Println("\nПисьмо от одногруппника не найдено.")
	} else {
		fmt.Println("\nПисьмо от одногруппника найдено. Сохраняем его...")
		seqset = new(imap.SeqSet)
		seqset.AddNum(uids[0])
		section := &imap.BodySectionName{}
		msgChan := make(chan *imap.Message, 1)
		go func() {
			if err := c.Fetch(seqset, []imap.FetchItem{section.FetchItem()}, msgChan); err != nil {
				log.Fatal(err)
			}
		}()

		msg := <-msgChan
		if msg == nil {
			fmt.Println("Не удалось получить письмо.")
		} else {
			r := msg.GetBody(section)
			if r == nil {
				fmt.Println("Тело письма пустое.")
			} else {
				f, err := os.Create("classmate_email.eml")
				if err != nil {
					return err
				}
				defer f.Close()
				if _, err := io.Copy(f, r); err != nil {
					return err
				}
				fmt.Println("Письмо от одногруппника сохранено в файл classmate_email.eml")
			}
		}
	}
	return nil
}

func main() {
	config, err := loadConfig("../config.json")
	if err != nil {
		log.Fatal("Ошибка загрузки конфигурации:", err)
	}

	if err := fetchEmails(config); err != nil {
		log.Fatal("Ошибка при получении писем:", err)
	}
}

File: 6/python/receiver_imap/main.py
import json
import imaplib
import email
import os
from email.header import decode_header

def load_config(filename="../../config.json"):
    with open(filename, "r", encoding="utf-8") as f:
        return json.load(f)

def decode_mime_words(s):
    if s:
        decoded_fragments = decode_header(s)
        return ''.join(
            fragment.decode(encoding if encoding else 'utf-8', errors='replace')
            if isinstance(fragment, bytes) else fragment
            for fragment, encoding in decoded_fragments
        )
    return s

def fetch_emails(config):
    imap_server = config["imapServer"]
    username = config["username"]
    password = config["password"]

    if ":" in imap_server:
        host, port = imap_server.split(":")
        port = int(port)
    else:
        host = imap_server
        port = 993

    mail_server = imaplib.IMAP4_SSL(host, port)
    mail_server.login(username, password)
    mail_server.select("INBOX")

    typ, data = mail_server.search(None, "ALL")
    if typ != "OK":
        print("Ошибка поиска писем")
        return

    msg_nums = data[0].split()
    print(f"Количество писем в ящике: {len(msg_nums)}")

    last_five = msg_nums[-5:] if len(msg_nums) >= 5 else msg_nums
    print("\nЗаголовки последних писем:")
    for num in last_five:
        typ, msg_data = mail_server.fetch(num, "(BODY.PEEK[HEADER])")
        if typ != "OK":
            print("Ошибка получения письма", num)
            continue
        msg = email.message_from_bytes(msg_data[0][1])
        from_header = decode_mime_words(msg.get("From"))
        subject_header = decode_mime_words(msg.get("Subject"))
        date_header = decode_mime_words(msg.get("Date"))
        print("От:", from_header, ", Тема:", subject_header, ", Дата:", date_header)

    search_criteria = f'(FROM "{config["to"]}")'
    typ, msg_nums = mail_server.search(None, search_criteria)
    if typ != "OK" or not msg_nums[0]:
        print("\nПисьмо от одногруппника не найдено.")
    else:
        msg_num = msg_nums[0].split()[0]
        typ, msg_data = mail_server.fetch(msg_num, "(RFC822)")
        if typ != "OK":
            print("Ошибка получения письма от одногруппника")
    mail_server.logout()

if __name__ == '__main__':
    config = load_config()
    fetch_emails(config)
